tsdb/head_append.go

```diff
@@ -1157,85 +1157,32 @@ func (s *memSeries) append(t int64, v float64, appendID uint64, o chunkOpts) (sa
-	// meta properly.
-	app, _ := s.app.(*chunkenc.HistogramAppender)
-	var (
-		pForwardInserts, nForwardInserts   []chunkenc.Insert
-		pBackwardInserts, nBackwardInserts []chunkenc.Insert
-		pMergedSpans, nMergedSpans         []histogram.Span
-		okToAppend, counterReset, gauge    bool
-	)
+	// chunk reference afterwards and mmap used up chunks.
+
+	// ignoring ok is ok, since we don't want to compare to the wrong previous appender anyway
+	prevApp, _ := s.app.(*chunkenc.HistogramAppender)
+
 	c, sampleInOrder, chunkCreated := s.appendPreprocessor(t, chunkenc.EncHistogram, o)
 	if !sampleInOrder {
 		return sampleInOrder, chunkCreated
 	}
-	switch h.CounterResetHint {
-	case histogram.GaugeType:
-		gauge = true
-		if app != nil {
-			pForwardInserts, nForwardInserts,
-				pBackwardInserts, nBackwardInserts,
-				pMergedSpans, nMergedSpans,
-				okToAppend = app.AppendableGauge(h)
-		}
-	case histogram.CounterReset:
-		// The caller tells us this is a counter reset, even if it
-		// doesn't look like one.
-		counterReset = true
-	default:
-		if app != nil {
-			pForwardInserts, nForwardInserts, okToAppend, counterReset = app.Appendable(h)
-		}
-	}
 
-	if !chunkCreated {
-		if len(pBackwardInserts)+len(nBackwardInserts) > 0 {
-			h.PositiveSpans = pMergedSpans
-			h.NegativeSpans = nMergedSpans
-			app.RecodeHistogram(h, pBackwardInserts, nBackwardInserts)
-		}
-		// We have 3 cases here
-		// - !okToAppend or counterReset -> We need to cut a new chunk.
-		// - okToAppend but we have inserts → Existing chunk needs
-		//   recoding before we can append our histogram.
-		// - okToAppend and no inserts → Chunk is ready to support our histogram.
-		switch {
-		case !okToAppend || counterReset:
-			c = s.cutNewHeadChunk(t, chunkenc.EncHistogram, o.chunkDiskMapper, o.chunkRange)
-			chunkCreated = true
-		case len(pForwardInserts) > 0 || len(nForwardInserts) > 0:
-			// New buckets have appeared. We need to recode all
-			// prior histogram samples within the chunk before we
-			// can process this one.
-			chunk, app := app.Recode(
-				pForwardInserts, nForwardInserts,
-				h.PositiveSpans, h.NegativeSpans,
-			)
-			c.chunk = chunk
-			s.app = app
-		}
-	}
+	var (
+		nc      chunkenc.Chunk
+		recoded bool
+		err     error
+	)
 
-	if chunkCreated {
-		hc := s.headChunk.chunk.(*chunkenc.HistogramChunk)
-		header := chunkenc.UnknownCounterReset
-		switch {
-		case gauge:
-			header = chunkenc.GaugeType
-		case counterReset:
-			header = chunkenc.CounterReset
-		case okToAppend:
-			header = chunkenc.NotCounterReset
-		}
-		hc.SetCounterResetHeader(header)
+	if !chunkCreated {
+		// ignore the previous appender if we continue the current chunk
+		prevApp = nil
 	}
 
-	s.app.AppendHistogram(t, h)
+	nc, recoded, s.app, err = s.app.AppendOrCreateHistogram(prevApp, t, h, false) // false=request a new chunk if needed
 
-	c.maxTime = t
+	if err != nil {
+		panic("appendOrCreateHistogram failed") // TODO: handle error
```

Comment:
This is somewhat related to #10305 , so we could first solve that by returning an error from this function and handling it properly.

Reply:
Question is if we could replace the `sampleInOrder bool` return value with an `error` efficiently. Where error=nil would be in order and error=storage.ErrOutOfOrderSample would mean out of order or some other error.

Reply:
Actually the only way we return error from AppanedHistogram here is if we failed to get the next Appender for a new empty chunk. Still the code would be cleaner if we had proper error handling.

Reply:
Having an error on an empty Appender is very unlikely, in fact [cutNewHeadChunk](https://github.com/prometheus/prometheus/blob/0e12f11d6112585415140bcdf0fd30c42e4255fe/tsdb/head_append.go#L1439) treats it as impossible and panics already.

Reply:
Created https://github.com/prometheus/prometheus/pull/12589 to check how it would like on main to have an error here.

Reply:
OK, then let's see where #12589 is going and address this later.

Reply:
I've reworked it a little to move the panic inside the AppendHistogram/AppendFloatHistogram functions where we can more easily argue about the fact that it will never happen.